src/protocols/bdx/BDXMessageUtils.cpp

```diff
@@ -0,0 +1,444 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      Implements utility methods for working with some complex BDX messages.
+ */
+
+#include <protocols/bdx/BDXMessageUtils.h>
+
+#include <support/BufBound.h>
+#include <support/BufferReader.h>
+#include <support/CodeUtils.h>
+
+#define VERSION_MASK 0x0F
+#define SENDER_DRIVE_MASK 0x10
+#define RECEIVER_DRIVE_MASK 0x20
+#define ASYNC_MASK 0x40
+#define CONTROL_MODE_MASK 0xF0
+
+#define DEFLEN_MASK 0x01
+#define START_OFFSET_MASK 0x02
+#define WIDERANGE_MASK 0x10
+
+using namespace chip;
+using namespace chip::BDX;
+using namespace chip::Encoding::LittleEndian;
+
+CHIP_ERROR TransferInit::Pack(System::PacketBuffer * aBuffer)
+{
+    CHIP_ERROR err   = CHIP_NO_ERROR;
+    uint8_t rangeCtl = 0, ptcByte = 0;
+
+    // BufBound has to be initialized with aBuffer so VerifyOrExit can't be used here.
+    if (aBuffer == NULL)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    BufBound bbuf(aBuffer->Start(), aBuffer->AvailableDataLength());
+
+    ptcByte |= mSupportedVersions & VERSION_MASK;
+    if (mSupportsSenderDrive)
+        ptcByte |= SENDER_DRIVE_MASK;
+    if (mSupportsReceiverDrive)
+        ptcByte |= RECEIVER_DRIVE_MASK;
+    if (mSupportsAsync)
+        ptcByte |= ASYNC_MASK;
+
+    bbuf.Put(ptcByte);
+    VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    if (mDefLen)
+        rangeCtl |= DEFLEN_MASK;
+    if (mStartOffset)
+        rangeCtl |= START_OFFSET_MASK;
+    if (mWideRange)
+        rangeCtl |= WIDERANGE_MASK;
+
+    bbuf.Put(rangeCtl);
+    VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    bbuf.PutLE16(mMaxBlockSize);
+    VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    if (mStartOffset > 0)
+    {
+        if (mWideRange)
+        {
+            bbuf.PutLE64(mStartOffset);
```

Comment:
since there are no consumers of the BE functions on bbuf, I suggest we remove them in favor of the semantics of BufReader, which assumes LE

Reply:
I suggest we leave the BE functions for use in implementing anything that may be IETF RFC-compliant since most RFC-specified protocols with data structures employ network byte order.

Reply:
how about we remove the LE suffix on the most commonly used ones?